src/pages/ExamsPage/hooks.js

```diff
@@ -28,11 +28,35 @@ export const useFetchCourseExams = () => {
   );
 };
 
+export const useFetchExamAttempts = () => {
+  const makeNetworkRequest = reduxHooks.useMakeNetworkRequest();
+  const dispatch = useDispatch();
+  return (examId) => (
+    makeNetworkRequest({
+      requestKey: RequestKeys.fetchExamAttempts,
+      promise: api.getExamAttempts(examId),
+      onSuccess: (response) => dispatch(reducer.loadExamAttempts(response)),
+    })
+  );
+};
+
 export const useInitializeExamsPage = (courseId) => {
   const fetchCourseExams = module.useFetchCourseExams();
   React.useEffect(() => { fetchCourseExams(courseId); }, []); // eslint-disable-line react-hooks/exhaustive-deps
 };
 
+// Return exam attempts for each exam in the course
+export const useInitializeAttemptsList = (exams) => {
+  const fetchExamAttempts = module.useFetchExamAttempts();
+  if (exams.length > 0) {
+    Object.values(exams).forEach((exam) => {
+      /* eslint-disable react-hooks/rules-of-hooks, react-hooks/exhaustive-deps */
+      React.useEffect(() => { fetchExamAttempts(exam.id); }, []);
```

Comment:
We should really only fetch attempts for the currently selected exam. So, we can trigger fetchExamAttempts on that changing either here or just in the component. (`useEffect() => { fetchExamAttempts(exam.id); }, [currentExam])`) TBH this is likely cleaner to do in ExamsPage/index.jsx and just pass the list to our list component like the example ExamList.jsx. Triggering the network request in AttemptList means this would re-fetch when you navigate between tabs.

The reason I have an initialize function with  a useEffect of `[]` is because that specific hook should run once on page load and never again. In this case we'd need to rerun each time the selected exam (currentExam) changes.

Also quick aside on useEffect and hooks in general. They must always run the same hooks in the same order on every re-render so doing a useEffect inside a forEach loop may cause react problems because it breaks the rules of hooks.

Reply:
Aaaah, yeah this makes a lot more sense. I'll be sure to implement this.